src/RequestParser.cpp

```diff
@@ -445,6 +469,69 @@ std::string	RequestParser::parseVersion(const std::string& requestLine)
 	return (requestLine.substr(++i));
 }
 
+
+void	RequestParser::parseChunkedBody(std::istringstream& requestStream)
+{
+	int			length = 0;
```

Comment:
It may be used to replace a value given by the header "Content-Length":

If a message is received with both a Transfer-Encoding and a Content-Length header field, the
Transfer-Encoding overrides the Content-Length. (RFC 9112, 6.3, p.19)

Question is: do we still do anything with that value afterwards in the response builder or elsewhere or can it be discarded?

Reply:
It seems like we could decide to error? 
Right after it says "Such a message might indicate an attempt to perform request smuggling (Section 11.2) or response splitting (Section 11.1) and ought to be handled as an error. "
Although it's not a MUST. An intermediary MUST scrub the length if he wants to forward the message.

A thought re program logic: the server config decides what the client max body size is. I think we would need to check beforehand if we reached this limit already (else a malicious client could bombard us with data). But the head does not count towards the body size.
So I think we would need to parse the header, see that it is chunked, leave and come back for the next batch?

Reply:
But they don´t specifiy what kind of error should be chosen. Probably 400, bad request again. 

What do you mean with the last sentence? 

The config might define a client request body size. Then we could check against this value with "length". But the config might not define it. Then we  would not need to count the length of the chunked body.  Also because we are not an intermediary which forwards the actual body length to an origin server.

Reply:
If the config does not define it, we should set a default. For nginx it's 1m: [Link](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size)

The problem is in the current implementation we read however long from the request stream. This indicates that we already received everything from the client and are now stitching together the body. This means a client could send us unlimited amount of data and we would happily take it. Then we would try in here to stitch it together - if we not already crashed, because no memory left.
What I'm trying to say is:
- we need to know that the client message is ready to be parsed, which we know by content-length or chunked encoding
- but we also need a way to come back here if we did not receive everything already (in case of chunked encoding)

Reply:
The way I see it we stay in this loop as long as not all the data (chunks) have been transmitted (indicated by reception of  the zero-chunk) or we realise that the length of the received data exceeds the given max size of the config file (or the default of 1m - I guess that's one megabyte). For that we would need to implement a check in the loop to verify this hard boundery has not been hit  yet. I would use the already exisiting issue #11 to track this. The issue itself would be rewritten.

Don´t see why we would leave the function though...

Reply:
Maybe I'm missing something.

RequestParser::parseHttpRequest receives a const std::string& request. It then checks if request line, headers and optional body are present (in this fixed order). 
This means RequestParser expects to receive the **full** request of the client. Therefore we have to receive and save as long as the client sends data. When the client doesn't send anymore only then can we go into the parser and check if this request is valid - because only then do we have the full request string.

I think we need to be able to partially check the request. The header gets parsed and the (first part of the) body. If our buffer can hold the full request we can parse it right away. If our buffer couldn't hold everything we need to validate the header and come back later when we received more. Or kick the client if he does stupid.

The server doesn't know if the request is valid since he can't parse the header. So he needs to accept everything

Reply:
I see now what you mean. 
The current implementation demands that the RequestParser object indeed receives the full request string, also when the body is chunked it would expect to receive the entire chunked body including the terminating zero-chunk. The aggregation of the chunked pieces happens beforehand in the handleConnections() function. 

I have added a check on whether the http request is complete - 
firstly checking for complete header section, then looking for whether a body is present differentiating between chunked and non-chunked bodies. Only when the body is completely transmitted, the parser is then called and receives the full string. Otherwise the handleConnections() function exits and is called again when data is available and adds the new data to the request string.

In order to avoid an endless acceptance of data (faulty requests w/o clear indicators of the body end) we will also check against the max. body size as discussed previously. This feature shall be introduced after the configParser is implemented.